src/backend/distributed/utils/colocation_utils.c

```diff
@@ -0,0 +1,189 @@
+ * out if given tables is not distributed.
+ */
+bool
+TablesColocated(Oid leftDistributedTableId, Oid rightDistributedTableId)
+{
+	int leftClocationId = TableColocationId(leftDistributedTableId);
+	int rightColocationId = TableColocationId(rightDistributedTableId);
+
+	if (leftClocationId == INVALID_COLOCATION_ID ||
+		rightColocationId == INVALID_COLOCATION_ID)
+	{
+		return false;
+	}
+
+	return leftClocationId == rightColocationId;
+}
+
+
+/*
+ * ShardsColocated function checks whether given two shards are co-located and
+ * returns true if they are co-located.
+ */
+bool
+ShardsColocated(ShardInterval *leftShardInterval, ShardInterval *rightShardInterval)
+{
+	bool tableColocated = TablesColocated(leftShardInterval->relationId,
+										  rightShardInterval->relationId);
+
+	if (tableColocated)
+	{
+		int leftShardMinValue = DatumGetInt32(leftShardInterval->minValue);
+		int leftShardMaxValue = DatumGetInt32(leftShardInterval->maxValue);
+		int rightShardMinValue = DatumGetInt32(rightShardInterval->minValue);
+		int rightShardMaxValue = DatumGetInt32(rightShardInterval->maxValue);
+
+		bool minValuesEqual = leftShardMinValue == rightShardMinValue;
+		bool maxValuesEqual = leftShardMaxValue == rightShardMaxValue;
+
+		return minValuesEqual && maxValuesEqual;
+	}
+
+	return false;
+}
+
+
+/*
+ * ColocatedTableList function returns list of distributedTableId which are co-located
+ * with given table.
+ */
+List *
+ColocatedTableList(Oid distributedTableId)
+{
+	List *colocatedTableList = NIL;
+	List *distributedTableList = DistTableOidList();
+	ListCell *distributedTableCell = NULL;
+
+	foreach(distributedTableCell, distributedTableList)
+	{
+		Oid currentDistributedTableId = lfirst_oid(distributedTableCell);
+		if (TablesColocated(distributedTableId, currentDistributedTableId))
+		{
+			colocatedTableList = lappend_oid(colocatedTableList,
+											 currentDistributedTableId);
+		}
+	}
+
+	return colocatedTableList;
+}
+
+
+/*
+ * ColocatedPlacementList function returns list of shardIntervals which are co-located
+ * with given shard.
+ */
+List *
+ColocatedShardList(ShardInterval *shardInterval)
+{
+	List *colocatedShardList = NIL;
+	Oid distributedTableId = shardInterval->relationId;
+	List *colocatedTableList = ColocatedTableList(distributedTableId);
+	ListCell *colocatedTableCell = NULL;
+
+	foreach(colocatedTableCell, colocatedTableList)
+	{
+		Oid colocatedTableId = lfirst_oid(colocatedTableCell);
+		List *shardIntervalList = LoadShardIntervalList(colocatedTableId);
+		ListCell *shardIntervalCell = NULL;
+		foreach(shardIntervalCell, shardIntervalList)
+		{
+			ShardInterval *currentShardInterval = (ShardInterval *) lfirst(
+				shardIntervalCell);
+			if (ShardsColocated(shardInterval, currentShardInterval))
+			{
+				colocatedTableList = lappend(colocatedShardList, currentShardInterval);
+
+				/*no need to search another co-located shard in this table */
+				break;
+			}
+		}
+	}
+
```

Comment:
This seems to be implemented in a rather expensive manner. The metadata caches are accessed at various levels, inside and outside of the loops.